### ISSUE TITLE
Mobius Addition Batch for Hyperboloid model?

### ISSUE THREAD

**Issue by AdarshMJ** [geoopt/geoopt#171]
Hi! 
So I have been trying to convert Hyperbolic Image Embeddings (https://arxiv.org/abs/1904.02239 ) from the Poincare ball model to the Lorentz model. The authors in the paper have used operations like -

`def _mobius_addition_batch(x, y, c):
    xy = _tensor_dot(x, y)  # B x C
    x2 = x.pow(2).sum(-1, keepdim=True)  # B x 1
    y2 = y.pow(2).sum(-1, keepdim=True)  # C x 1
    num = 1 + 2 * c * xy + c * y2.permute(1, 0)  # B x C
    num = num.unsqueeze(2) * x.unsqueeze(1)
    num = num + (1 - c * x2).unsqueeze(2) * y  # B x C x D
    denom_part1 = 1 + 2 * c * xy  # B x C
    denom_part2 = c ** 2 * x2 * y2.permute(1, 0)
    denom = denom_part1 + denom_part2
    res = num / (denom.unsqueeze(2) + 1e-5)
    return res`

`def _hyperbolic_softmax(X, A, P, c):
    lambda_pkc = 2 / (1 - c * P.pow(2).sum(dim=1))
    k = lambda_pkc * torch.norm(A, dim=1) / torch.sqrt(c)
    mob_add = _mobius_addition_batch(-P, X, c)
    num = 2 * torch.sqrt(c) * torch.sum(mob_add * A.unsqueeze(1), dim=-1)
    denom = torch.norm(A, dim=1, keepdim=True) * (1 - c * mob_add.pow(2).sum(dim=2))
    logit = k.unsqueeze(1) * arsinh(num / denom)
    return logit.permute(1, 0)`

I am not entirely sure what are the hyperboloid model equivalents of these are. My strategy was to use the same operations and then map the points to the hyperboloid model using the formula -

1+||w||^2/2*||w|*(|1-||w||). (from https://arxiv.org/pdf/1903.07307.pdf section 2.2)

But doesn't seem to be working. Any insights are welcome please!
